### ISSUE TITLE
`\path ... foreach ...` doesn't work with option `turn`

### ISSUE THREAD

**Issue by muzimuzhi** [pgf-tikz/pgf#1047]
**Brief outline of the bug**

from TeX-SX question: https://tex.stackexchange.com/questions/615588/coordinate-turn-option-in-foreach

In short, value of `\tikz@tangent` is not smuggled outside the TeX-group added to each execution of foreach loop body.
https://github.com/pgf-tikz/pgf/blob/85d2c38b78407e5b1ba42140dfd11fcdf8df1edf/tex/generic/pgf/frontendlayer/tikz/tikz.code.tex#L371-L384

Is saving/restoring `\tikz@tangant` in pgffor hooks, as what did for `\tikz@last[xy]` the recommended fixing approach?
https://github.com/pgf-tikz/pgf/blob/85d2c38b78407e5b1ba42140dfd11fcdf8df1edf/tex/generic/pgf/frontendlayer/tikz/tikz.code.tex#L2582-L2616

**Minimal working example (MWE)**

```latex
\documentclass{article}
\usepackage{tikz}

\begin{document}
Expected:
\begin{tikzpicture}
  \draw[red] (0,0) -- (1,.2) -- ([turn]90:1cm) -- ([turn]90:1cm) -- ([turn]90:1cm) -- ([turn]90:1cm);
\end{tikzpicture}

Actual:
\begin{tikzpicture}
  \draw (0,0) -- (1,.2) foreach \i in {1,2,3,4} { -- ([turn]90:1cm) };
\end{tikzpicture}
\end{document}
```
![image](https://user-images.githubusercontent.com/6376638/133629291-84bf57c4-55fc-4f2d-a3f2-c78a1f663993.png)

**Comment by hmenke** [comment]
Adding it to the hooks is not very elegant, but definitely much more elegant than just making `\tikz@tangent` global. Please go ahead.

**Comment by muzimuzhi** [comment]
Or we could add sth like `\pgffor@smuggleone` or `\aftergroup\pgffor@smuggled` to `pgffor`, since I guess there will be more in tikz which need to be smuggled, e.g., `\tikz@timer` (not sure). It seems the new idea can avoid assigning things alternatively local and global.

**Comment by rumout** [comment]
Hello,

Any update on this bug? I found myself caught on this bug today.

As a workaround, I placed the code above on the preamble without issues. However, I'm just loading tikz package without any others (other packages, libraries and such), so I'm concerned if this would raise some issue on other pgf/tikz parts...

Thanks!

**Comment by muzimuzhi** [comment]
> As a workaround, I placed the code above on the preamble without issues. [...], so I'm concerned if this would raise some issue on other pgf/tikz parts...

@rumout If you're talking about the patch I provided in this [TeX.SX answer](https://tex.stackexchange.com/a/615651), no I think it won't raise other issues.

**Comment by rumout** [comment]
@muzimuzhi yes, I was talking about that... Any chances you will be pushing this changes into "main" code?...
